subsys/bluetooth/services/ras/rrsp/ras_rrsp.c

```diff
@@ -8,21 +8,684 @@
+		struct bt_ras_rrsp *rrsp = &rrsp_pool[i];
+
+		if (rrsp->conn == conn) {
+			return rrsp;
+		}
+	}
+
+	return NULL;
+}
+
+int bt_ras_rrsp_alloc(struct bt_conn *conn)
+{
+	struct bt_ras_rrsp *rrsp = NULL;
+
+	if (rrsp_find(conn) != NULL) {
+		return -EALREADY;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(rrsp_pool); i++) {
+		if (rrsp_pool[i].conn == NULL) {
+			rrsp = &rrsp_pool[i];
+			break;
+		}
+	}
+
+	if (rrsp == NULL) {
+		return -ENOMEM;
+	}
+
+	LOG_DBG("conn %p new rrsp %p", (void *)conn, (void *)rrsp);
+
+	memset(rrsp, 0, sizeof(struct bt_ras_rrsp));
+	rrsp->conn = bt_conn_ref(conn);
+
+	k_work_init(&rrsp->send_data_work, send_data_work_handler);
+	k_work_init(&rrsp->rascp_work, rascp_work_handler);
+	k_work_init(&rrsp->status_work, status_work_handler);
+	k_timer_init(&rrsp->rascp_timeout, rascp_timeout_handler, NULL);
+
+	return 0;
+}
+
+void bt_ras_rrsp_free(struct bt_conn *conn)
+{
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	if (rrsp) {
+		LOG_DBG("conn %p rrsp %p", (void *)rrsp->conn, (void *)rrsp);
+
+		(void)k_work_cancel(&rrsp->send_data_work);
+		(void)k_work_cancel(&rrsp->rascp_work);
+		(void)k_work_cancel(&rrsp->status_work);
+		k_timer_stop(&rrsp->rascp_timeout);
+
+		k_work_queue_drain(&rrsp_wq, false);
+
+		bt_conn_unref(rrsp->conn);
+		rrsp->conn = NULL;
+	}
+}
+
+static ssize_t ras_features_read(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				 uint16_t len, uint16_t offset)
 {
 	return bt_gatt_attr_read(conn, attr, buf, len, offset, &ras_features, sizeof(ras_features));
 }
 
+static ssize_t rd_ready_read(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			     uint16_t len, uint16_t offset)
+{
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	if (!rrsp) {
+		return BT_GATT_ERR(BT_ATT_ERR_READ_NOT_PERMITTED);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &rrsp->ready_ranging_counter,
+				 sizeof(uint16_t));
+}
+
+static ssize_t rd_overwritten_read(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				   uint16_t len, uint16_t offset)
+{
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	if (!rrsp) {
+		return BT_GATT_ERR(BT_ATT_ERR_READ_NOT_PERMITTED);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &rrsp->overwritten_ranging_counter,
+				 sizeof(uint16_t));
+}
+
+static ssize_t ras_cp_write(struct bt_conn *conn, struct bt_gatt_attr const *attr, void const *buf,
+			    uint16_t len, uint16_t offset, uint8_t flags)
+{
+	LOG_HEXDUMP_DBG(buf, len, "Write request");
+
+	if (!bt_gatt_is_subscribed(conn, attr, BT_GATT_CCC_INDICATE)) {
+		LOG_DBG("Not subscribed");
+		return BT_GATT_ERR(RAS_ATT_ERROR_CCC_CONFIG);
+	}
+
+	struct bt_ras_rrsp *rrsp = rrsp_find(conn);
+
+	if (!rrsp || k_work_is_pending(&rrsp->rascp_work) || len > RASCP_WRITE_MAX_LEN) {
+		LOG_DBG("Write rejected");
+		return BT_GATT_ERR(RAS_ATT_ERROR_WRITE_REQ_REJECTED);
+	}
+
+	uint16_t buf_len = MIN(len, RASCP_WRITE_MAX_LEN);
```

Comment:
Do we still need this? If `len` is more than `RASCP_WRITE_MAX_LEN` we will return above

Reply:
can remove the variable entirely

Reply:
Good point, will remove